pkg/util/util.go

```diff
@@ -149,70 +160,160 @@ func CloneRepo(clonePath, repoURL string) error {
-			devfileBytes, err = ioutil.ReadFile(path.Join(localpath, f.Name()))
+			devfileBytes, err := ioutil.ReadFile(path.Join(localpath, f.Name()))
 			if err != nil {
-				return nil, err
+				return nil, nil, err
 			}
 
-			var context string
-			currentPath := localpath
-			for i := 0; i < currentLevel; i++ {
-				context = path.Join(filepath.Base(currentPath), context)
-				currentPath = filepath.Dir(currentPath)
-			}
+			context := getContext(localpath, currentLevel)
 			devfileMapFromRepo[context] = devfileBytes
+			isDevfilePresent = true
 		} else if f.IsDir() && f.Name() == devfile.HiddenDevfileDir {
 			// if the dir is .devfile, we dont increment currentLevel
 			// consider devfile.yaml and .devfile/devfile.yaml as the same level
-			recursiveMap, err := searchDevfiles(path.Join(localpath, f.Name()), currentLevel, depth)
+			recursiveDevfileMap, recursiveDevfileURLMap, err := searchDevfiles(path.Join(localpath, f.Name()), currentLevel, depth)
 			if err != nil {
-				return nil, err
+				return nil, nil, err
 			}
 
-			var context string
-			currentPath := localpath
-			for i := 0; i < currentLevel; i++ {
-				context = path.Join(filepath.Base(currentPath), context)
-				currentPath = filepath.Dir(currentPath)
-			}
-			for recursiveContext := range recursiveMap {
+			context := getContext(localpath, currentLevel)
+			for recursiveContext := range recursiveDevfileMap {
 				if recursiveContext == devfile.HiddenDevfileDir {
-					devfileMapFromRepo[context] = recursiveMap[devfile.HiddenDevfileDir]
+					devfileMapFromRepo[context] = recursiveDevfileMap[devfile.HiddenDevfileDir]
+					devfilesURLMapFromRepo[context] = recursiveDevfileURLMap[devfile.HiddenDevfileDir]
+					isDevfilePresent = true
 				}
 			}
 		} else if f.IsDir() {
 			if currentLevel+1 <= depth {
-				recursiveMap, err := searchDevfiles(path.Join(localpath, f.Name()), currentLevel+1, depth)
+				recursiveDevfileMap, recursiveDevfileURLMap, err := searchDevfiles(path.Join(localpath, f.Name()), currentLevel+1, depth)
 				if err != nil {
-					return nil, err
+					return nil, nil, err
 				}
-				for context, devfile := range recursiveMap {
+				for context, devfile := range recursiveDevfileMap {
 					devfileMapFromRepo[context] = devfile
+					devfilesURLMapFromRepo[context] = recursiveDevfileURLMap[context]
+					isDevfilePresent = true
 				}
 			}
 		}
 	}
 
-	// if we didnt find any devfile we should return an err
 	if len(devfileMapFromRepo) == 0 && currentLevel == 0 {
-		err = fmt.Errorf("unable to find any devfile(s) in the multi component repo, devfiles can be detected only upto a depth of %v dir", depth)
+		// if we didnt find any devfile we should return an err
+		err = &NoDevfileFound{location: localpath}
+	} else if !isDevfilePresent && currentLevel == depth {
+		// if we didnt find any devfile upto our desired depth, then use alizer
+		devfileBytes, detectedDevfileEndpoint, err := AnalyzeAndDetectDevfile(localpath)
```

Comment:
If we only ever call alizer at the max depth (I’m assuming that’s what localpath is set to), do we risk missing things if we needed to have scanned at a higher depth (if that makes sense)